File: tools/golden-gen/src/golden_gen/release_transport.py
"""Schema5 byte transport; installation alone never asserts release acceptance."""

from __future__ import annotations

import gzip
import hashlib
import json
import os
import shutil
import tarfile
import tempfile
import zlib
from collections.abc import Generator
from pathlib import Path
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from golden_gen.assets import _rename_no_replace
from golden_gen.layered_artifacts import sha

MANIFEST_LIMIT = 16 * 1024**2
ARTIFACT_LIMIT = 50_000
FILE_LIMIT = 8 * 1024**3 - 1
TOTAL_LIMIT = 1024**4
# GitHub's documented per-asset limit, checked 2026-09-08: strictly under 2 GiB.
UPLOAD_LIMIT = 2 * 1024**3
ARCHIVE = "goldens-v0.2.tar.gz"
ASSETS = {"manifest.json", ARCHIVE}
HEX64 = r"^[0-9a-f]{64}$"
HEX40 = r"^[0-9a-f]{40}$"


def logical_path(value: str) -> str:
    if (
        not value
        or not value.isascii()
        or len(value) > 512
        or "\\" in value
        or "\0" in value
        or any(p in ("", ".", "..", ".git") for p in value.split("/"))
    ):
        raise ValueError("unsafe logical artifact path")
    return value


def regular(path: Path) -> None:
    if path.is_symlink() or not path.is_file():
        raise ValueError("artifact is not a regular file")


def no_symlinks(path: Path) -> None:
    if any(p.is_symlink() for p in (path, *path.parents)):
        raise ValueError("symlink in artifact or cache path")


class Source(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, strict=True)
    commit: str = Field(pattern=HEX40)
    tree: str = Field(pattern=HEX40)


class Artifact(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, strict=True)
    logical_path: str
    filename: str = Field(pattern=r"^artifact-[0-9]{6}\.bin$")
    sha256: str = Field(pattern=HEX64)
    size_bytes: int = Field(ge=0, le=FILE_LIMIT)

    @model_validator(mode="after")
    def safe(self) -> Artifact:
        logical_path(self.logical_path)
        return self

    @classmethod
    def inventory(cls, root: Path, paths: list[Path]) -> list[Artifact]:
        relative = sorted({p.relative_to(root).as_posix() for p in paths})
        result = []
        for index, name in enumerate(relative):
            logical_path(name)
            path = root / name
            regular(path)
            if any(p.is_symlink() for p in path.parents if p != root.parent):
                raise ValueError("symlink in artifact path")
            result.append(
                cls(
                    logical_path=name,
                    filename=f"artifact-{index:06}.bin",
                    sha256=sha(path),
                    size_bytes=path.stat().st_size,
                )
            )
        return result


class ArchiveIdentity(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, strict=True)
    filename: Literal["goldens-v0.2.tar.gz"]
    sha256: str = Field(pattern=HEX64)


class Entrypoints(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, strict=True)
    authoritative_manifest: str
    authoritative_marker: str
    performance: str
    cpu_gates: str


class ReleaseManifest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, strict=True)
    schema_version: Literal[5] = 5
    protocol: Literal["layered-accuracy-v1"] = "layered-accuracy-v1"
    product_version: Literal["v0.2.0"] = "v0.2.0"
    golden_version: Literal["goldens-v0.2"] = "goldens-v0.2"
    source: Source
    registry_sha256: str = Field(pattern=HEX64)
    policy_sha256: str = Field(pattern=HEX64)
    definition_index_blob: str = Field(pattern=HEX40)
    entrypoints: Entrypoints
    counts: dict[str, int]
    artifacts: list[Artifact] = Field(min_length=1, max_length=ARTIFACT_LIMIT)
    archive: ArchiveIdentity | None = None

    @model_validator(mode="after")
    def exact_inventory(self) -> ReleaseManifest:
        names = [a.logical_path for a in self.artifacts]
        if names != sorted(set(names)):
            raise ValueError("logical inventory must be sorted and unique")
        files = set(names)
        for index, item in enumerate(self.artifacts):
            if item.filename != f"artifact-{index:06}.bin":
                raise ValueError("noncanonical flat inventory filename")
            if any(
                "/".join(item.logical_path.split("/")[:n]) in files
                for n in range(1, len(item.logical_path.split("/")))
            ):
                raise ValueError("file/directory logical path conflict")
        if sum(a.size_bytes for a in self.artifacts) > TOTAL_LIMIT:
            raise ValueError("extracted size limit exceeded")
        if any(p not in files for p in self.entrypoints.model_dump().values()):
            raise ValueError("missing release entrypoint")
        if any(type(n) is not int or not 0 <= n <= 2**64 - 1 for n in self.counts.values()):
            raise ValueError("invalid release counts")
        return self


def read_manifest(path: Path) -> ReleaseManifest:
    regular(path)
    if path.stat().st_size > MANIFEST_LIMIT:
        raise ValueError("release manifest exceeds size limit")
    # Required wire keys cannot be supplied by producer-side model defaults.
    with path.open("rb") as stream:
        raw = stream.read(MANIFEST_LIMIT + 1)
    if len(raw) > MANIFEST_LIMIT:
        raise ValueError("release manifest exceeds size limit")

    def unique(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
        result: dict[str, Any] = {}
        for key, value in pairs:
            if key in result:
                raise ValueError("duplicate release manifest key")
            result[key] = value
        return result

    data = json.loads(raw, object_pairs_hook=unique)
    if not isinstance(data, dict) or set(data) != set(ReleaseManifest.model_fields):
        raise ValueError("release manifest missing or unknown wire fields")
    if type(data["schema_version"]) is not int:
        raise ValueError("release schema_version must be a wire integer")
    value = ReleaseManifest.model_validate(data)
    if value.archive is None:
        raise ValueError("release manifest has no archive identity")
    return value


def check_upload_size(size: int) -> None:
    if size >= UPLOAD_LIMIT:
        raise ValueError("capacity decision required: GitHub asset must be under 2 GiB")


def _capacity(parent: Path, size: int) -> None:
    if shutil.disk_usage(parent).free < size + MANIFEST_LIMIT:
        raise ValueError("capacity decision required: insufficient disk space")


def _header(item: Artifact) -> bytes:
    info = tarfile.TarInfo(item.filename)
    info.size = item.size_bytes
    info.mode = 0o644
    return info.tobuf(format=tarfile.USTAR_FORMAT)


def _verify_file(path: Path, item: Artifact) -> None:
    regular(path)
    if path.stat().st_size != item.size_bytes or sha(path) != item.sha256:
        raise ValueError("artifact size/checksum mismatch")


def build_bundle(root: Path, manifest: ReleaseManifest, destination: Path) -> Path:
    """Produce transport bytes only; the release adapter must first validate semantics."""
    no_symlinks(root)
    no_symlinks(destination.parent)
    if destination.exists() or destination.is_symlink():
        raise FileExistsError(destination)
    _capacity(destination.parent, 2 * sum(a.size_bytes + 1024 for a in manifest.artifacts))
    staging = Path(tempfile.mkdtemp(prefix=".schema5-", dir=destination.parent))
    try:
        archive = staging / ARCHIVE
        with archive.open("xb") as raw:
            with gzip.GzipFile(filename="", mode="wb", fileobj=raw, mtime=0) as compressed:
                for item in manifest.artifacts:
                    path = root / item.logical_path
                    no_symlinks(path)
                    if not path.resolve().is_relative_to(root.resolve()):
                        raise ValueError("artifact escapes evidence root")
                    _verify_file(path, item)
                    compressed.write(_header(item))
                    copied = 0
                    digest = hashlib.sha256()
                    with path.open("rb") as source:
                        while chunk := source.read(64 * 1024):
                            copied += len(chunk)
                            if copied > item.size_bytes:
                                raise ValueError("artifact grew during archive streaming")
                            digest.update(chunk)
                            compressed.write(chunk)
                    if copied != item.size_bytes or digest.hexdigest() != item.sha256:
                        raise ValueError("artifact changed during archive streaming")
                    compressed.write(bytes((-item.size_bytes) % 512))
                    _verify_file(path, item)
                compressed.write(bytes(1024))
            raw.flush()
            os.fsync(raw.fileno())
        check_upload_size(archive.stat().st_size)
        bound = manifest.model_copy(
            update={"archive": ArchiveIdentity(filename="goldens-v0.2.tar.gz", sha256=sha(archive))}
        )
        payload = (bound.model_dump_json(indent=2) + "\n").encode()
        if len(payload) > MANIFEST_LIMIT:
            raise ValueError("release manifest exceeds size limit")
        (staging / "manifest.json").write_bytes(payload)
        _rename_no_replace(staging, destination)
        return destination
    finally:
        if staging.exists():
            shutil.rmtree(staging)


def _inflate(path: Path) -> Generator[bytes, None, None]:
    decoder = zlib.decompressobj(31)
    with path.open("rb") as stream:
        if stream.read(10) != bytes.fromhex("1f8b08000000000002ff"):
            raise ValueError("noncanonical gzip header")
        stream.seek(0)
        pending = b""
        while not decoder.eof:
            pending = pending or stream.read(64 * 1024)
            if not pending:
                raise ValueError("truncated gzip stream")
            output = decoder.decompress(pending, 64 * 1024)
            pending = decoder.unconsumed_tail
            yield output
        if decoder.unused_data or pending or stream.read(1):
            raise ValueError("trailing compressed content")


class _Reader:
    def __init__(self, path: Path):
        self.chunks = _inflate(path)
        self.buffer = b""

    def read(self, size: int) -> bytes:
        while len(self.buffer) < size:
            chunk = next(self.chunks, None)
            if chunk is None:
                break
            self.buffer += chunk
        result, self.buffer = self.buffer[:size], self.buffer[size:]
        return result


def verify_install(installed: Path, manifest_bytes: bytes, manifest: ReleaseManifest) -> None:
    no_symlinks(installed)
    regular(installed / "manifest.json")
    if (installed / "manifest.json").read_bytes() != manifest_bytes:
        raise ValueError("installed manifest differs")
    expected = {"manifest.json", *("evidence/" + a.logical_path for a in manifest.artifacts)}
    found = set()
    directories = {"evidence"}
    for artifact in manifest.artifacts:
        directories.update(
            p.as_posix()
            for p in Path("evidence", artifact.logical_path).parents
            if p.as_posix() != "."
        )
    for path in installed.rglob("*"):
        if path.is_symlink():
            raise ValueError("symlink in immutable install")
        if path.is_file():
            found.add(path.relative_to(installed).as_posix())
        elif path.is_dir() and path.relative_to(installed).as_posix() not in directories:
            raise ValueError("unexpected directory in immutable install")
        elif not path.is_dir():
            raise ValueError("special entry in immutable install")
    if found != expected:
        raise ValueError("immutable inventory mismatch")
    for item in manifest.artifacts:
        _verify_file(installed / "evidence" / item.logical_path, item)


def install_transport(bundle: Path, cache: Path) -> Path:
    """Strict transport verification; caller must separately evaluate release semantics."""
    no_symlinks(bundle)
    no_symlinks(cache)
    if {p.name for p in bundle.iterdir()} != ASSETS:
        raise ValueError("release must contain exactly two assets")
    manifest = read_manifest(bundle / "manifest.json")
    assert manifest.archive is not None
    archive = bundle / ARCHIVE
    regular(archive)
    check_upload_size(archive.stat().st_size)
    if sha(archive) != manifest.archive.sha256:
        raise ValueError("archive checksum mismatch")
    parent = cache / "goldens-v0.2"
    parent.mkdir(parents=True, exist_ok=True)
    if parent.is_symlink() or cache.is_symlink():
        raise ValueError("symlink cache root")
    final = parent / manifest.archive.sha256
    payload = (bundle / "manifest.json").read_bytes()
    if final.exists() or final.is_symlink():
        if final.is_symlink():
            raise ValueError("symlink immutable install")
        verify_install(final, payload, manifest)
        return final
    _capacity(parent, sum(a.size_bytes for a in manifest.artifacts))
    staging = Path(tempfile.mkdtemp(prefix=".install-", dir=parent))
    reader = _Reader(archive)
    try:
        for item in manifest.artifacts:
            if reader.read(512) != _header(item):
                raise ValueError("noncanonical or unexpected USTAR header")
            target = staging / "evidence" / item.logical_path
            target.parent.mkdir(parents=True, exist_ok=True)
            digest = hashlib.sha256()
            remaining = item.size_bytes
            with target.open("xb") as stream:
                while remaining:
                    chunk = reader.read(min(remaining, 64 * 1024))
                    if not chunk:
                        raise ValueError("truncated archive payload")
                    remaining -= len(chunk)
                    digest.update(chunk)
                    stream.write(chunk)
            if digest.hexdigest() != item.sha256:
                raise ValueError("extracted artifact checksum mismatch")
            padding = (-item.size_bytes) % 512
            if reader.read(padding) != bytes(padding):
                raise ValueError("nonzero USTAR payload padding")
        tail = reader.read(1024 * 1024 + 1)
        if len(tail) < 1024 or len(tail) > 1024 * 1024 or len(tail) % 512 or any(tail):
            raise ValueError("invalid USTAR end markers or trailing content")
        (staging / "manifest.json").write_bytes(payload)
        verify_install(staging, payload, manifest)
        try:
            _rename_no_replace(staging, final)
        except FileExistsError:
            verify_install(final, payload, manifest)
        return final
    finally:
        reader.chunks.close()
        if staging.exists():
            shutil.rmtree(staging)

File: crates/vllm-oxide-test/src/release_transport.rs
//! Schema5 transport only. Semantic release acceptance is a separate operation.

use std::collections::{BTreeMap, BTreeSet};
use std::fs::{File, OpenOptions};
use std::io::{BufReader, Read, Write};
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};

use anyhow::{bail, Context, Result};
use flate2::bufread::GzDecoder;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

const MANIFEST_LIMIT: u64 = 16 * 1024 * 1024;
const FILE_LIMIT: u64 = 8 * 1024 * 1024 * 1024 - 1;
const TOTAL_LIMIT: u64 = 1024 * 1024 * 1024 * 1024;
const UPLOAD_LIMIT: u64 = 2 * 1024 * 1024 * 1024;
const ARCHIVE: &str = "goldens-v0.2.tar.gz";
static NONCE: AtomicU64 = AtomicU64::new(0);

#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Source {
    pub commit: String,
    pub tree: String,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Artifact {
    pub logical_path: String,
    pub filename: String,
    pub sha256: String,
    pub size_bytes: u64,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ArchiveIdentity {
    pub filename: String,
    pub sha256: String,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Entrypoints {
    pub authoritative_manifest: String,
    pub authoritative_marker: String,
    pub performance: String,
    pub cpu_gates: String,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct ReleaseManifest {
    pub schema_version: u32,
    pub protocol: String,
    pub product_version: String,
    pub golden_version: String,
    pub source: Source,
    pub registry_sha256: String,
    pub policy_sha256: String,
    pub definition_index_blob: String,
    pub entrypoints: Entrypoints,
    #[serde(deserialize_with = "unique_counts")]
    pub counts: BTreeMap<String, u64>,
    pub artifacts: Vec<Artifact>,
    pub archive: ArchiveIdentity,
}

fn unique_counts<'de, D: serde::Deserializer<'de>>(
    deserializer: D,
) -> std::result::Result<BTreeMap<String, u64>, D::Error> {
    struct Unique;
    impl<'de> serde::de::Visitor<'de> for Unique {
        type Value = BTreeMap<String, u64>;
        fn expecting(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.write_str("unique count names")
        }
        fn visit_map<M: serde::de::MapAccess<'de>>(
            self,
            mut map: M,
        ) -> std::result::Result<Self::Value, M::Error> {
            let mut values = BTreeMap::new();
            while let Some((key, value)) = map.next_entry::<String, u64>()? {
                if values.insert(key, value).is_some() {
                    return Err(serde::de::Error::custom("duplicate count name"));
                }
            }
            Ok(values)
        }
    }
    deserializer.deserialize_map(Unique)
}

fn no_symlinks(path: &Path) -> Result<()> {
    for ancestor in path.ancestors() {
        match std::fs::symlink_metadata(ancestor) {
            Ok(m) if m.file_type().is_symlink() => bail!("symlink in artifact or cache path"),
            Ok(_) => {}
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {}
            Err(e) => return Err(e.into()),
        }
    }
    Ok(())
}

fn hex(value: &str, size: usize) -> bool {
    value.len() == size
        && value
            .bytes()
            .all(|v| v.is_ascii_digit() || (b'a'..=b'f').contains(&v))
}

fn safe_path(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 512
        && value.is_ascii()
        && !value.contains(['\\', '\0'])
        && value
            .split('/')
            .all(|p| !matches!(p, "" | "." | ".." | ".git"))
}

pub fn parse_manifest(bytes: &[u8]) -> Result<ReleaseManifest> {
    if u64::try_from(bytes.len())? > MANIFEST_LIMIT {
        bail!("release manifest exceeds size limit");
    }
    let value: ReleaseManifest = serde_json::from_slice(bytes)?;
    if value.schema_version != 5
        || value.protocol != "layered-accuracy-v1"
        || value.product_version != "v0.2.0"
        || value.golden_version != "goldens-v0.2"
        || value.archive.filename != ARCHIVE
        || !hex(&value.archive.sha256, 64)
        || !hex(&value.source.commit, 40)
        || !hex(&value.source.tree, 40)
        || !hex(&value.registry_sha256, 64)
        || !hex(&value.policy_sha256, 64)
        || !hex(&value.definition_index_blob, 40)
        || value.artifacts.is_empty()
        || value.artifacts.len() > 50_000
    {
        bail!("invalid schema5 release identity or bounds");
    }
    let mut names = BTreeSet::new();
    let mut previous: Option<&str> = None;
    let mut total = 0_u64;
    for (index, artifact) in value.artifacts.iter().enumerate() {
        if !safe_path(&artifact.logical_path)
            || previous.is_some_and(|p| p >= artifact.logical_path.as_str())
            || artifact.filename != format!("artifact-{index:06}.bin")
            || !hex(&artifact.sha256, 64)
            || artifact.size_bytes > FILE_LIMIT
        {
            bail!("invalid artifact inventory");
        }
        let mut prefix = String::new();
        for part in artifact.logical_path.split('/') {
            if names.contains(&prefix) {
                bail!("file/directory path conflict");
            }
            if !prefix.is_empty() {
                prefix.push('/');
            }
            prefix.push_str(part);
        }
        names.insert(artifact.logical_path.clone());
        previous = Some(&artifact.logical_path);
        total = total
            .checked_add(artifact.size_bytes)
            .context("size overflow")?;
    }
    if total > TOTAL_LIMIT {
        bail!("extracted size limit exceeded");
    }
    for path in [
        &value.entrypoints.authoritative_manifest,
        &value.entrypoints.authoritative_marker,
        &value.entrypoints.performance,
        &value.entrypoints.cpu_gates,
    ] {
        if !names.contains(path) {
            bail!("missing release entrypoint");
        }
    }
    Ok(value)
}

fn regular(path: &Path) -> Result<()> {
    if !std::fs::symlink_metadata(path)?.file_type().is_file() {
        bail!("not a regular artifact");
    }
    Ok(())
}

fn digest(path: &Path) -> Result<String> {
    regular(path)?;
    let mut stream = File::open(path)?;
    let mut hash = Sha256::new();
    let mut buffer = [0_u8; 65_536];
    loop {
        let count = stream.read(&mut buffer)?;
        if count == 0 {
            break;
        }
        hash.update(&buffer[..count]);
    }
    Ok(format!("{:x}", hash.finalize()))
}

fn header(artifact: &Artifact) -> [u8; 512] {
    let mut bytes = [0_u8; 512];
    bytes[..artifact.filename.len()].copy_from_slice(artifact.filename.as_bytes());
    for (start, len, number) in [
        (100, 8, 0o644),
        (108, 8, 0),
        (116, 8, 0),
        (124, 12, artifact.size_bytes),
        (136, 12, 0),
    ] {
        let octal = format!("{number:0width$o}\0", width = len - 1);
        bytes[start..start + len].copy_from_slice(octal.as_bytes());
    }
    bytes[148..156].fill(b' ');
    bytes[156] = b'0';
    bytes[257..263].copy_from_slice(b"ustar\0");
    bytes[263..265].copy_from_slice(b"00");
    let checksum: u64 = bytes.iter().map(|v| u64::from(*v)).sum();
    bytes[148..156].copy_from_slice(format!("{checksum:06o}\0 ").as_bytes());
    bytes
}

fn verify_tree(root: &Path, bytes: &[u8], manifest: &ReleaseManifest) -> Result<()> {
    no_symlinks(root)?;
    regular(&root.join("manifest.json"))?;
    if std::fs::read(root.join("manifest.json"))? != bytes {
        bail!("installed manifest differs");
    }
    let expected = std::iter::once("manifest.json".to_string())
        .chain(
            manifest
                .artifacts
                .iter()
                .map(|a| format!("evidence/{}", a.logical_path)),
        )
        .collect::<BTreeSet<_>>();
    let mut actual = BTreeSet::new();
    let mut directories = BTreeSet::new();
    for artifact in &manifest.artifacts {
        let path = PathBuf::from("evidence").join(&artifact.logical_path);
        for p in path.ancestors().skip(1) {
            if !p.as_os_str().is_empty() {
                directories.insert(p.to_path_buf());
            }
        }
    }
    let mut pending = vec![root.to_path_buf()];
    while let Some(directory) = pending.pop() {
        for entry in std::fs::read_dir(directory)? {
            let entry = entry?;
            if entry.file_type()?.is_dir() {
                if !directories.contains(entry.path().strip_prefix(root)?) {
                    bail!("unexpected installed directory");
                }
                pending.push(entry.path());
            } else if entry.file_type()?.is_file() {
                actual.insert(
                    entry
                        .path()
                        .strip_prefix(root)?
                        .to_str()
                        .context("non-UTF8 path")?
                        .to_string(),
                );
            } else {
                bail!("symlink or special installed entry");
            }
        }
    }
    if actual != expected {
        bail!("immutable inventory mismatch");
    }
    for artifact in &manifest.artifacts {
        let path = root.join("evidence").join(&artifact.logical_path);
        if path.metadata()?.len() != artifact.size_bytes || digest(&path)? != artifact.sha256 {
            bail!("installed size/checksum mismatch");
        }
    }
    Ok(())
}

struct Staging(PathBuf);
impl Drop for Staging {
    fn drop(&mut self) {
        let _ = std::fs::remove_dir_all(&self.0);
    }
}

pub fn install_transport(bundle: &Path, cache: &Path) -> Result<PathBuf> {
    no_symlinks(bundle)?;
    no_symlinks(cache)?;
    let names = std::fs::read_dir(bundle)?
        .map(|e| Ok(e?.file_name()))
        .collect::<Result<BTreeSet<_>>>()?;
    if names
        != [
            std::ffi::OsString::from("manifest.json"),
            std::ffi::OsString::from(ARCHIVE),
        ]
        .into()
    {
        bail!("release must contain exactly two assets");
    }
    let path = bundle.join("manifest.json");
    regular(&path)?;
    if path.metadata()?.len() > MANIFEST_LIMIT {
        bail!("manifest size limit");
    }
    let mut bytes = Vec::new();
    File::open(path)?
        .take(MANIFEST_LIMIT + 1)
        .read_to_end(&mut bytes)?;
    let manifest = parse_manifest(&bytes)?;
    let archive = bundle.join(ARCHIVE);
    regular(&archive)?;
    if archive.metadata()?.len() >= UPLOAD_LIMIT || digest(&archive)? != manifest.archive.sha256 {
        bail!("archive size/checksum mismatch");
    }
    let parent = cache.join("goldens-v0.2");
    std::fs::create_dir_all(&parent)?;
    if cache.symlink_metadata()?.file_type().is_symlink()
        || parent.symlink_metadata()?.file_type().is_symlink()
    {
        bail!("symlink cache root");
    }
    let final_path = parent.join(&manifest.archive.sha256);
    if let Ok(metadata) = std::fs::symlink_metadata(&final_path) {
        if !metadata.is_dir() {
            bail!("invalid immutable install");
        }
        verify_tree(&final_path, &bytes, &manifest)?;
        return Ok(final_path);
    }
    let disk = rustix::fs::statvfs(&parent)?;
    let needed = manifest.artifacts.iter().map(|a| a.size_bytes).sum::<u64>() + MANIFEST_LIMIT;
    if disk.f_bavail.saturating_mul(disk.f_frsize) < needed {
        bail!("capacity decision required");
    }
    let mut stage = None;
    for _ in 0..1024 {
        let path = parent.join(format!(
            ".schema5-{}-{}",
            std::process::id(),
            NONCE.fetch_add(1, Ordering::Relaxed)
        ));
        match std::fs::create_dir(&path) {
            Ok(()) => {
                stage = Some(Staging(path));
                break;
            }
            Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => {}
            Err(e) => return Err(e.into()),
        }
    }
    let stage = stage.context("cannot allocate staging directory")?;
    let mut compressed = BufReader::new(File::open(archive)?);
    let mut gzip_header = [0_u8; 10];
    compressed.read_exact(&mut gzip_header)?;
    if gzip_header != [31, 139, 8, 0, 0, 0, 0, 0, 2, 255] {
        bail!("noncanonical gzip header");
    }
    use std::io::Seek;
    compressed.rewind()?;
    let mut reader = GzDecoder::new(compressed);
    let mut buffer = [0_u8; 65_536];
    for artifact in &manifest.artifacts {
        let mut block = [0_u8; 512];
        reader.read_exact(&mut block)?;
        if block != header(artifact) {
            bail!("noncanonical or unexpected USTAR header");
        }
        let target = stage.0.join("evidence").join(&artifact.logical_path);
        std::fs::create_dir_all(target.parent().context("missing parent")?)?;
        let mut output = OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(target)?;
        let mut remaining = artifact.size_bytes;
        let mut hash = Sha256::new();
        while remaining > 0 {
            let count = reader.read(&mut buffer[..usize::try_from(remaining.min(65_536))?])?;
            if count == 0 {
                bail!("truncated archive payload");
            }
            output.write_all(&buffer[..count])?;
            hash.update(&buffer[..count]);
            remaining -= u64::try_from(count)?;
        }
        output.sync_all()?;
        if format!("{:x}", hash.finalize()) != artifact.sha256 {
            bail!("extracted checksum mismatch");
        }
        let padding = usize::try_from((512 - artifact.size_bytes % 512) % 512)?;
        reader.read_exact(&mut buffer[..padding])?;
        if buffer[..padding].iter().any(|v| *v != 0) {
            bail!("nonzero payload padding");
        }
    }
    let mut tail = Vec::new();
    reader
        .by_ref()
        .take(1024 * 1024 + 1)
        .read_to_end(&mut tail)?;
    if tail.len() < 1024
        || tail.len() > 1024 * 1024
        || tail.len() % 512 != 0
        || tail.iter().any(|v| *v != 0)
    {
        bail!("invalid USTAR end markers or trailing content");
    }
    if reader.into_inner().read(&mut buffer[..1])? != 0 {
        bail!("trailing compressed content");
    }
    std::fs::write(stage.0.join("manifest.json"), &bytes)?;
    verify_tree(&stage.0, &bytes, &manifest)?;
    match rustix::fs::renameat_with(
        rustix::fs::CWD,
        &stage.0,
        rustix::fs::CWD,
        &final_path,
        rustix::fs::RenameFlags::NOREPLACE,
    ) {
        Ok(()) => {}
        Err(rustix::io::Errno::EXIST) => verify_tree(&final_path, &bytes, &manifest)?,
        Err(e) => return Err(e.into()),
    }
    Ok(final_path)
}
